code/linalg.c

```diff
@@ -274,16 +274,16 @@ mp_obj_t linalg_eye(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args)
     ndarray_obj_t *ndarray = create_new_ndarray(m, n, dtype);
     mp_obj_t one = mp_obj_new_int(1);
     size_t i = 0;
-    if((k >= 0) && (k < n)) {
-        while(k < n) {
+    if((k >= 0) && (abs(k) < n)) {
```

Comment:
If k >= 0, then abs(k) = k, isn't it? I don't see what the absolute value is supposed to do here...

Reply:
This is probably to fix a signed vs unsigned comparison warning, since CircuitPython has `-Wall` set. Currently `n` is not validated as > 0, which it probably should be. After validation, an intermediate `validated_n` that is a `uint32_t` or `size_t` could be used instead.

Reply:
is there any issue with using abs()?

Reply:
Using `abs()` is tricking things up, when the true fix is to make sure `n` is >=  zero and then cast it appropriately to an unsigned. As I mentioned, the code should be validating `n`, since a negative value will cause havoc, I think.

Reply:
since n is unsigned, wouldn't we want to create an intermediate validated_k that is unsigned if k>=0?

Reply:
Fine with me.

Reply:
Well, I guess, both `n` and `k` need validation, since the code does:
```c
    size_t n = args[0].u_int, m;
    int16_t k = args[2].u_int;
```
if `n` is negative, that will cause it to be a very large positive value. And `k` could be negative, but `int16_t` could also end up truncating `k`. I'm not familiar with the API to know what the restrictions on these args are. @v923z, are you avoiding arg validation for speed reasons? I imagine there are other places where there is code like this.

Reply:
I don't think that we should be concerned with speed here: what you suggest is a single `if` statement, or something similar, while the code is meant to iterate over arrays, and apply whatever mathematical functions to them. That is gonna be much more expensive than just validating the argument at the beginning. So, if we need to do that, then let's do that.